// @libp2p/prometheus-metrics/dist/src/counter.js
import { Counter as PromCounter } from 'prom-client';
import { normaliseString } from './utils.js';
export class PrometheusCounter {
    constructor(name, opts) {
        name = normaliseString(name);
        const help = normaliseString(opts.help ?? name);
        const labels = opts.label != null ? [normaliseString(opts.label)] : [];
        let collect;
        this.calculators = [];
        // calculated metric
        if (opts?.calculate != null) {
            this.calculators.push(opts.calculate);
            const self = this;
            collect = async function () {
                const values = await Promise.all(self.calculators.map(async (calculate) => await calculate()));
                const sum = values.reduce((acc, curr) => acc + curr, 0);
                this.inc(sum);
            };
        }
        this.counter = new PromCounter({
            name,
            help,
            labelNames: labels,
            registers: opts.registry !== undefined ? [opts.registry] : undefined,
            collect
        });
    }
    addCalculator(calculator) {
        this.calculators.push(calculator);
    }
    increment(value = 1) {
        this.counter.inc(value);
    }
    reset() {
        this.counter.reset();
    }
}
