// @libp2p/kad-dht/dist/src/topology-listener.js
import { createTopology } from '@libp2p/topology';
import { CustomEvent, EventEmitter } from '@libp2p/interfaces/events';
import { logger } from '@libp2p/logger';
/**
 * Receives notifications of new peers joining the network that support the DHT protocol
 */
export class TopologyListener extends EventEmitter {
    constructor(components, init) {
        super();
        const { protocol, lan } = init;
        this.components = components;
        this.log = logger(`libp2p:kad-dht:topology-listener:${lan ? 'lan' : 'wan'}`);
        this.running = false;
        this.protocol = protocol;
    }
    isStarted() {
        return this.running;
    }
    /**
     * Start the network
     */
    async start() {
        if (this.running) {
            return;
        }
        this.running = true;
        // register protocol with topology
        const topology = createTopology({
            onConnect: (peerId) => {
                this.log('observed peer %p with protocol %s', peerId, this.protocol);
                this.dispatchEvent(new CustomEvent('peer', {
                    detail: peerId
                }));
            }
        });
        this.registrarId = await this.components.registrar.register(this.protocol, topology);
    }
    /**
     * Stop all network activity
     */
    async stop() {
        this.running = false;
        // unregister protocol and handlers
        if (this.registrarId != null) {
            this.components.registrar.unregister(this.registrarId);
            this.registrarId = undefined;
        }
    }
}
